compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -3390,4 +3390,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             err.span_label(block.span, "this block is missing a tail expression");
         }
     }
+
+    pub(crate) fn suggest_swapping_lhs_and_rhs(
+        &self,
+        err: &mut Diag<'_>,
+        checked_ty: Ty<'tcx>,
+        expected_ty_expr: &'tcx hir::Expr<'tcx>,
+        op: hir::BinOp,
+    ) {
+        match op.node {
+            hir::BinOpKind::Eq => {
+                if let Some(partial_eq_def_id) = self.infcx.tcx.lang_items().eq_trait()
+                    && let lhs_ty = self.check_expr(expected_ty_expr)
+                    && self
+                        .infcx
+                        .type_implements_trait(
+                            partial_eq_def_id,
+                            [checked_ty, lhs_ty],
+                            self.param_env,
+                        )
+                        .must_apply_modulo_regions()
+                {
+                    let trait_name = self.infcx.tcx.item_name(partial_eq_def_id);
+                    err.note(format!("`{}` implements `{}<{}>`", checked_ty, trait_name, lhs_ty));
+                    err.help(format!("consider swapping the LHS and RHS of the equality"));
```

Comment:
This could be a [suggestion](https://rustc-dev-guide.rust-lang.org/diagnostics.html#suggestions)